File: kai/psa/erpnext.py
"""
ERPNext ITSM Bridge for KAI

Provides customer portal integration, billing, and full ITSM features.
"""

import httpx
import logging
from typing import Dict, List, Optional, Any
from datetime import datetime

logger = logging.getLogger(__name__)


class ERPNextClient:
    """ERPNext REST API client for ITSM and customer portal integration"""
    
    def __init__(
        self,
        base_url: str,
        api_key: str,
        api_secret: str,
        timeout: int = 30
    ):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.api_secret = api_secret
        self.timeout = timeout
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            headers={
                "Authorization": f"token {api_key}:{api_secret}",
                "Content-Type": "application/json"
            },
            timeout=timeout
        )
    
    async def create_issue(
        self,
        subject: str,
        description: str,
        customer: str,
        priority: str = "Medium",
        issue_type: str = "Bug",
        **kwargs
    ) -> Dict[str, Any]:
        """
        Create issue in ERPNext
        
        Args:
            subject: Issue title
            description: Issue details
            customer: Customer ID
            priority: Low/Medium/High/Critical
            issue_type: Bug/Feature/Question/Incident
            
        Returns:
            Created issue document
        """
        data = {
            "doctype": "Issue",
            "subject": subject,
            "description": description,
            "customer": customer,
            "priority": priority,
            "issue_type": issue_type,
            "status": "Open",
            "raised_by": kwargs.get("raised_by", ""),
            **kwargs
        }
        
        response = await self.client.post("/api/resource/Issue", json=data)
        response.raise_for_status()
        return response.json()["data"]
    
    async def get_issue(self, issue_id: str) -> Dict[str, Any]:
        """Get issue by ID"""
        response = await self.client.get(f"/api/resource/Issue/{issue_id}")
        response.raise_for_status()
        return response.json()["data"]
    
    async def update_issue(
        self,
        issue_id: str,
        **updates
    ) -> Dict[str, Any]:
        """Update issue fields"""
        response = await self.client.put(
            f"/api/resource/Issue/{issue_id}",
            json=updates
        )
        response.raise_for_status()
        return response.json()["data"]
    
    async def add_comment(
        self,
        issue_id: str,
        comment: str,
        comment_by: str
    ) -> Dict[str, Any]:
        """Add comment to issue"""
        data = {
            "doctype": "Comment",
            "comment_type": "Comment",
            "reference_doctype": "Issue",
            "reference_name": issue_id,
            "content": comment,
            "comment_by": comment_by
        }
        
        response = await self.client.post("/api/resource/Comment", json=data)
        response.raise_for_status()
        return response.json()["data"]
    
    async def get_customer_issues(
        self,
        customer: str,
        status: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """Get all issues for a customer"""
        filters = [["customer", "=", customer]]
        if status:
            filters.append(["status", "=", status])
        
        response = await self.client.get(
            "/api/resource/Issue",
            params={
                "filters": str(filters),
                "fields": '["name", "subject", "status", "priority", "creation"]'
            }
        )
        response.raise_for_status()
        return response.json()["data"]
    
    async def create_customer_portal_user(
        self,
        email: str,
        first_name: str,
        last_name: str,
        customer: str
    ) -> Dict[str, Any]:
        """Create customer portal user"""
        data = {
            "doctype": "Contact",
            "email_id": email,
            "first_name": first_name,
            "last_name": last_name,
            "links": [{
                "link_doctype": "Customer",
                "link_name": customer
            }]
        }
        
        response = await self.client.post("/api/resource/Contact", json=data)
        response.raise_for_status()
        return response.json()["data"]
    
    async def get_customer_contracts(
        self,
        customer: str
    ) -> List[Dict[str, Any]]:
        """Get active contracts for customer"""
        filters = [
            ["party_name", "=", customer],
            ["status", "=", "Active"]
        ]
        
        response = await self.client.get(
            "/api/resource/Contract",
            params={
                "filters": str(filters),
                "fields": '["name", "contract_terms", "start_date", "end_date"]'
            }
        )
        response.raise_for_status()
        return response.json()["data"]
    
    async def get_customer_assets(
        self,
        customer: str
    ) -> List[Dict[str, Any]]:
        """Get assets for customer"""
        filters = [["customer", "=", customer]]
        
        response = await self.client.get(
            "/api/resource/Asset",
            params={
                "filters": str(filters),
                "fields": '["name", "asset_name", "item_code", "status"]'
            }
        )
        response.raise_for_status()
        return response.json()["data"]
    
    async def create_sales_invoice(
        self,
        customer: str,
        items: List[Dict[str, Any]],
        **kwargs
    ) -> Dict[str, Any]:
        """Create sales invoice (billing integration)"""
        data = {
            "doctype": "Sales Invoice",
            "customer": customer,
            "items": items,
            "posting_date": datetime.now().strftime("%Y-%m-%d"),
            **kwargs
        }
        
        response = await self.client.post("/api/resource/Sales Invoice", json=data)
        response.raise_for_status()
        return response.json()["data"]
    
    async def get_knowledge_base_articles(
        self,
        category: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """Get knowledge base articles"""
        filters = []
        if category:
            filters.append(["category", "=", category])
        
        response = await self.client.get(
            "/api/resource/Help Article",
            params={
                "filters": str(filters) if filters else "[]",
                "fields": '["name", "title", "content", "category"]'
            }
        )
        response.raise_for_status()
        return response.json()["data"]
    
    async def close(self):
        """Close HTTP client"""
        await self.client.aclose()

File: internal/psa/erpnext.go
// Package erpnext provides ERPNext REST API client for billing and ITSM integration
package erpnext

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Client is an ERPNext API client
type Client struct {
	BaseURL   string
	APIKey    string
	APISecret string
	HTTPClient *http.Client
}

// NewClient creates a new ERPNext client
func NewClient(baseURL, apiKey, apiSecret string) *Client {
	return &Client{
		BaseURL:   baseURL,
		APIKey:    apiKey,
		APISecret: apiSecret,
		HTTPClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// Issue represents an ERPNext issue/ticket
type Issue struct {
	Name        string    `json:"name,omitempty"`
	Subject     string    `json:"subject"`
	Description string    `json:"description"`
	Customer    string    `json:"customer"`
	Priority    string    `json:"priority"` // Low, Medium, High, Critical
	IssueType   string    `json:"issue_type"` // Bug, Feature, Question, Incident
	Status      string    `json:"status,omitempty"`
	RaisedBy    string    `json:"raised_by,omitempty"`
	Creation    time.Time `json:"creation,omitempty"`
}

// SalesInvoice represents an ERPNext sales invoice
type SalesInvoice struct {
	Name        string              `json:"name,omitempty"`
	Customer    string              `json:"customer"`
	PostingDate string              `json:"posting_date"`
	Items       []SalesInvoiceItem  `json:"items"`
	Status      string              `json:"status,omitempty"`
}

// SalesInvoiceItem represents a line item
type SalesInvoiceItem struct {
	ItemCode    string  `json:"item_code"`
	ItemName    string  `json:"item_name"`
	Description string  `json:"description,omitempty"`
	Qty         float64 `json:"qty"`
	Rate        float64 `json:"rate"`
	Amount      float64 `json:"amount"`
}

// CreateIssue creates a new issue in ERPNext
func (c *Client) CreateIssue(ctx context.Context, issue *Issue) (*Issue, error) {
	issue.Status = "Open"
	
	data := map[string]interface{}{
		"doctype":     "Issue",
		"subject":     issue.Subject,
		"description": issue.Description,
		"customer":    issue.Customer,
		"priority":    issue.Priority,
		"issue_type":  issue.IssueType,
		"status":      issue.Status,
		"raised_by":   issue.RaisedBy,
	}
	
	var result Issue
	if err := c.doRequest(ctx, "POST", "/api/resource/Issue", data, &result); err != nil {
		return nil, fmt.Errorf("create issue: %w", err)
	}
	
	return &result, nil
}

// GetIssue retrieves an issue by ID
func (c *Client) GetIssue(ctx context.Context, issueID string) (*Issue, error) {
	var result Issue
	if err := c.doRequest(ctx, "GET", fmt.Sprintf("/api/resource/Issue/%s", issueID), nil, &result); err != nil {
		return nil, fmt.Errorf("get issue: %w", err)
	}
	
	return &result, nil
}

// UpdateIssue updates an issue
func (c *Client) UpdateIssue(ctx context.Context, issueID string, updates map[string]interface{}) (*Issue, error) {
	var result Issue
	if err := c.doRequest(ctx, "PUT", fmt.Sprintf("/api/resource/Issue/%s", issueID), updates, &result); err != nil {
		return nil, fmt.Errorf("update issue: %w", err)
	}
	
	return &result, nil
}

// CreateSalesInvoice creates a sales invoice for billing
func (c *Client) CreateSalesInvoice(ctx context.Context, invoice *SalesInvoice) (*SalesInvoice, error) {
	if invoice.PostingDate == "" {
		invoice.PostingDate = time.Now().Format("2006-01-02")
	}
	
	data := map[string]interface{}{
		"doctype":      "Sales Invoice",
		"customer":     invoice.Customer,
		"posting_date": invoice.PostingDate,
		"items":        invoice.Items,
	}
	
	var result SalesInvoice
	if err := c.doRequest(ctx, "POST", "/api/resource/Sales Invoice", data, &result); err != nil {
		return nil, fmt.Errorf("create sales invoice: %w", err)
	}
	
	return &result, nil
}

// doRequest performs an HTTP request to ERPNext API
func (c *Client) doRequest(ctx context.Context, method, path string, body interface{}, result interface{}) error {
	url := c.BaseURL + path
	
	var reqBody []byte
	var err error
	if body != nil {
		reqBody, err = json.Marshal(body)
		if err != nil {
			return fmt.Errorf("marshal request: %w", err)
		}
	}
	
	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewReader(reqBody))
	if err != nil {
		return fmt.Errorf("create request: %w", err)
	}
	
	req.Header.Set("Authorization", fmt.Sprintf("token %s:%s", c.APIKey, c.APISecret))
	req.Header.Set("Content-Type", "application/json")
	
	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return fmt.Errorf("do request: %w", err)
	}
	defer resp.Body.Close()
	
	if resp.StatusCode >= 400 {
		return fmt.Errorf("ERPNext API error: %s", resp.Status)
	}
	
	if result != nil {
		var apiResp struct {
			Data json.RawMessage `json:"data"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
			return fmt.Errorf("decode response: %w", err)
		}
		
		if err := json.Unmarshal(apiResp.Data, result); err != nil {
			return fmt.Errorf("unmarshal data: %w", err)
		}
	}
	
	return nil
}
